### PR TITLE
zarmeen's model and associations

### PR DESCRIPTION
**Pull request by zarmeenmustafa651**

### CODE REVIEW

File: db/migrate/20200908065226_create_lab_reports.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class CreateLabReports < ActiveRecord::Migration[6.0]
+  def change
+    create_table :lab_reports do |t|
+      t.references :patient, null: false, foreign_key: true
+      t.references :test, null: false, foreign_key: true
+      t.datetime :sample_collection_date
+      t.datetime :report_generation_date
+
+      t.timestamps
```
Comment (itratzahra7) [inline db/migrate/20200908065226_create_lab_reports.rb]: null: false

File: test/models/availability_test.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
Comment (MirahSaqib) [inline test/models/availability_test.rb]: remove these test files, we will use rspec for test cases.

File: app/models/medicine.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+class Medicine < ApplicationRecord
+	has_many :bill_details as: :billable, dependent: :nullify
+	has_many :prescribed_medicines, dependent: :nullify
+	has_many :purchase_details, dependent: :nullify
+    belongs_to :hospital
```
Comment (MirahSaqib) [inline app/models/medicine.rb]: fix spacing and alignment

File: app/models/prescribed_medicine.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+class PrescribedMedicine < ApplicationRecord
+  belongs_to :appointment
+  belongs_to :medicine
+  belongs_to :hospital
+
```
Comment (MirahSaqib) [inline app/models/prescribed_medicine.rb]: run rubocop, remove extra empty lines.

File: db/migrate/20200908065226_create_lab_reports.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class CreateLabReports < ActiveRecord::Migration[6.0]
+  def change
+    create_table :lab_reports do |t|
+      t.references :patient, null: false, foreign_key: true
+      t.references :hospital, null: false, foreign_key: true
+      t.references :test, null: false, foreign_key: true
+      t.datetime :sample_collection_date, null: false
+      t.datetime :report_generation_date, null: false
+
+      t.timestamps
```
Comment (MirahSaqib) [inline db/migrate/20200908065226_create_lab_reports.rb]: add null: false with timestamps

File: db/migrate/20200908065232_create_medicines.rb [hunk 6]
```diff
@@ -0,0 +1,12 @@
+class CreateMedicines < ActiveRecord::Migration[6.0]
+  def change
+    create_table :medicines do |t|
+      t.string :name, null: false
+      t.integer :quantity, null: false
+      t.decimal :price, precision: 10, scale: 2, null: false
+      t.references :hospital, null: false, foreign_key: true
```
Comment (MirahSaqib) [inline db/migrate/20200908065232_create_medicines.rb:7]: Add composite index on ```hospital_id``` and  ```medicine_id```.

File: db/migrate/20200908093226_create_prescribed_medicines.rb [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+  def change
+    create_table :prescribed_medicines do |t|
+      t.references :appointment, null: false, foreign_key: true
+      t.references :medicine, null: false, foreign_key: true
+      t.string :usage_instruction, null: false
+      t.references :hospital, null: false, foreign_key: true
+
+      t.timestamps
+    end
+  end
```
Comment (MirahSaqib) [inline db/migrate/20200908093226_create_prescribed_medicines.rb:13]: add index on ```medicine_id```

File: db/migrate/20200908065226_create_lab_reports.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    end
+  end
```
Comment (MirahSaqib) [inline db/migrate/20200908065226_create_lab_reports.rb:13]: Add composite index on ```hospital_id``` and ```lab_report_id```
do for other migrations as well, add index on hospital_id in all tables

File: app/models/test.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+class Test < ApplicationRecord
+	has_many :bill_details as: :billable, dependent: :nullify
+	has_many :lab_reports, dependent: :nullify
+	has_many :bills, through: :bill_details
```
Comment (MirahSaqib) [inline app/models/test.rb:4]: fix spacing and alignment

**Review by MirahSaqib — COMMENTED** [review]
Cycle 1

File: db/migrate/20200908065226_create_lab_reports.rb (hunk 8 above)
Comment (zarmeenmustafa651) [reply db/migrate/20200908065226_create_lab_reports.rb:13]: should I add a composite index in all tables or index on the hospital_id only?

File: db/migrate/20200908093226_create_prescribed_medicines.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    create_table :prescribed_medicines do |t|
+      t.references :appointment, null: false, foreign_key: true
+      t.references :medicine, null: false, foreign_key: true
+      t.string :usage_instruction, null: false
+      t.references :hospital, null: false, foreign_key: true
+
+      t.timestamps null: false
+    end
+    add_index :prescribed_medicines, :hospital
+    add_index :prescribed_medicines, :medicine
```
Comment (MirahSaqib) [inline db/migrate/20200908093226_create_prescribed_medicines.rb]: medicine_id

File: db/migrate/20200908093226_create_prescribed_medicines.rb [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+  def change
+    create_table :prescribed_medicines do |t|
+      t.references :appointment, null: false, foreign_key: true
+      t.references :medicine, null: false, foreign_key: true
+      t.string :usage_instruction, null: false
+      t.references :hospital, null: false, foreign_key: true
+
+      t.timestamps null: false
+    end
+    add_index :prescribed_medicines, :hospital
```
Comment (MirahSaqib) [inline db/migrate/20200908093226_create_prescribed_medicines.rb]: hospital_id

File: db/migrate/20200908065232_create_medicines.rb [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+      t.timestamps null: false
+    end
+    add_index :medicines, [:hospital, :id]
```
Comment (MirahSaqib) [inline db/migrate/20200908065232_create_medicines.rb]: hospital_id

File: db/migrate/20200908065229_create_availabilities.rb [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+      t.string :week_day, null: false
+      t.datetime :start_slot, null: false
+      t.datetime :end_slot, null: false
+      t.references :doctor, null: false, foreign_key: true
+      t.references :hospital, null: false, foreign_key: true
+
+      t.timestamps null: false
+    end
+    add_index :availabilities, :week_day
+    add_index :availabilities, :hospital
```
Comment (MirahSaqib) [inline db/migrate/20200908065229_create_availabilities.rb]: hospital_id

File: db/migrate/20200908065228_create_tests.rb [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+    create_table :tests do |t|
+      t.string :name, null: false
+      t.string :details, null: false
+      t.decimal :price, precision: 10, scale: 2
+      t.references :hospital, null: false, foreign_key: true
+
+      t.timestamps null: false
+    end
+    add_index :tests, :name
+    add_index :tests, :hospital
```
Comment (MirahSaqib) [inline db/migrate/20200908065228_create_tests.rb]: hospital_id

File: db/migrate/20200908065226_create_lab_reports.rb [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+    create_table :lab_reports do |t|
+      t.references :patient, null: false, foreign_key: true
+      t.references :hospital, null: false, foreign_key: true
+      t.references :test, null: false, foreign_key: true
+      t.datetime :sample_collection_date, null: false
+      t.datetime :report_generation_date, null: false
+
+      t.timestamps null: false
+    end
+    add_index :lab_reports, [:hospital, :id]
```
Comment (MirahSaqib) [inline db/migrate/20200908065226_create_lab_reports.rb]: hospital_id

**Pull request merged**